feat: produce es, common, and umd bundles

AndyOGo opened the pull request:
fixes #106 

@JustinBeckwith 
This PR utilizes [`microbundle`](https://github.com/developit/microbundle) to publish ressources for:
- ES Modules
- CommonJS
- and UMD

This enhances compatibility with module bundlers and loaders like rollup, parcel or webpack.

Running `npm run build`, results in:

```sh
npm run build

> retry-axios@0.0.0 build /Users/axawinterthur/dev/retry-axios
> microbundle --tsconfig=tsconfig.json

Build "retryAxios" to dist:
        819 B: index.js.gz
        725 B: index.js.br
        805 B: index.module.js.gz
        717 B: index.module.js.br
        899 B: index.umd.js.gz
        812 B: index.umd.js.br
```

**Note:** I moved the original `tsconfig.json` to `tsconfig.compile.json` to leave the test setup unchanged.

AndyOGo commented on package.json:
```diff
@@ -2,8 +2,11 @@
-  "types": "./build/src/index.d.ts",
+  "source": "src/index.ts",
+  "main": "dist/index.js",
+  "module": "dist/index.module.js",
```
About `package.module`:

> `pkg.module` will point to a module that has ES2015 module syntax but **otherwise only syntax features that the target environments support**.
https://github.com/rollup/rollup/wiki/pkg.module

> Wait, it just means import and export – not other future JavaScript features?
> **Yes. ...**
https://github.com/rollup/rollup/wiki/pkg.module#wait-it-just-means-import-and-export--not-other-future-javascript-features

AndyOGo commented on tsconfig.json:
```diff
@@ -1,11 +1,10 @@
   "include": [
-    "src/*.ts",
-    "test/*.ts"
+    "src/*.ts"
```
For distribution only `src` is required

JustinBeckwith commented on .gitignore:
```diff
@@ -4,4 +4,8 @@ umd/
 coverage/
 .nyc_output/
 .vscode/
+.rts2_cache_cjs/
```
Maybe we should do `.rts2*` and replace all these?

AndyOGo replied:
Done

JustinBeckwith commented on .npmignore:
```diff
@@ -0,0 +1,11 @@
+node_modules
```
We use the `files` property in `package.json`, which produces an *opt-in* list of files we want to ship.  As of today, that's only the contents of `build/src` (along with defaults like `package.json`, `README.md`, etc).  You probably don't need this at all!

AndyOGo replied:
Great, I didn't see that, will remove `.npmigore` :)

AndyOGo replied:
Done

JustinBeckwith commented on package.json:
```diff
@@ -2,8 +2,11 @@
   "description": "Retry HTTP requests with Axios.",
-  "main": "./build/src/index.js",
-  "types": "./build/src/index.d.ts",
+  "source": "src/index.ts",
```
What is `source`?

AndyOGo replied:
It tells microbundle where to find the source file to bundle.

> Unless overridden via the command line, microbundle uses the `source` property in your `package.json` to locate the input file, and the `main` property for the output.
https://github.com/developit/microbundle#microbundle--microbundle-build

JustinBeckwith replied:
Ok, so this is only consumed at build time?  Not by downstream consumers?  I ask because the `src` directory is *not* included in the npm module.

AndyOGo replied:
Correct, this runs during the NPM's prepare phase.

Yes, I killed `build/src`.
Okay, so is `src` needed by downstream consumers?
With this change I replaced `build/src` with `dist` and kept `build/*` for running tests.

So far consumers received this tarball via npm:
https://registry.npmjs.org/retry-axios/-/retry-axios-2.1.3.tgz

AndyOGo replied:
@JustinBeckwith If you prefer it not to be declarative as a source field.
There is also a cli option available for `--entry` (`-i`)
https://github.com/developit/microbundle#all-cli-options

JustinBeckwith replied:
Nah `source` is fine, I just wanted to make sure none of the tools here required it at runtime (as opposed to build time).

JustinBeckwith commented on package.json:
```diff
@@ -2,8 +2,11 @@
-  "main": "./build/src/index.js",
-  "types": "./build/src/index.d.ts",
+  "source": "src/index.ts",
+  "main": "dist/index.js",
```
Need to make sure to add `dist/` to the `files` array below

AndyOGo replied:
done

JustinBeckwith commented on package.json:
```diff
@@ -16,10 +19,11 @@
     "codecov": "c8 report --reporter=json && codecov -f coverage/*.json",
     "fix": "gts fix",
-    "compile": "tsc -p .",
+    "compile": "tsc -p tsconfig.compile.json",
```
Having a `compile` and a `build` command feels wildly confusing to me.  Can you describe how these are different, and in which situations they're used?

AndyOGo replied:
The build produces 3 formats, and I think these aren't necessary for the tests.
For this reason I kept them separated, but I agree it's not really DRY 🤔

JustinBeckwith replied:
I think I'm ok with just changing the target for TypeScript to `es2015` if that helps keep this simpler?  Additionally, maybe `build-web` would be a better name?

JustinBeckwith commented on tsconfig.compile.json:
```diff
@@ -0,0 +1,11 @@
+{
```
Is the idea here to produce different sets of compiled sources for es2018 and es2015 environments?  Just trying to keep up with what's happening :)

AndyOGo replied:
This is for running tests in node only.

JustinBeckwith replied:
If possible, let's just re-target the whole library to es2015

AndyOGo commented:
@JustinBeckwith Thank you very much for your quick response and your great review.
I addressed your feedback a rebased onto your updated master branch.

Let me know what you think.

JustinBeckwith commented on package.json:
```diff
@@ -47,13 +51,14 @@
     "typescript": "~3.8.0"
   },
   "files": [
-    "build/src"
```
... we don't need both?

AndyOGo replied:
No, as far as I'm aware of `dist/` is enough.

The `prepare` script bundles only into `dist/`.
And `compile` is only execute at `pretest` and compiles into `build/`.

AndyOGo commented:
Hello @JustinBeckwith 
How are you?
Did you have a chance to look an my changes regarding your feedback?
Im looking forward to your feedback.

AndyOGo commented:
Hmh, ci fails with 🤔 
> Error: Failed to load plugin '@typescript-eslint' declared in '.eslintrc.json » ./node_modules/gts#overrides[0]': Cannot find module 'json-schema'

JustinBeckwith commented on tsconfig.json:
```diff
@@ -2,7 +2,10 @@
+    "outDir": "dist",
+    "moduleResolution": "node",
+    "module": "ES2015",
+    "allowSyntheticDefaultImports": true
```
I would prefer to keep `allowSyntheticDefaultImports` disabled unless we *really* need it.  It's far too easy to break downstream modules with that flag set.

AndyOGo replied:
Okay.
hmh, without it calling `assert` and `nock` did not work anymore.
I have to check again, if I can make it work with that flag 🤔

AndyOGo replied:
done

AndyOGo commented on package.json:
```diff
@@ -16,10 +19,11 @@
     "codecov": "c8 report --reporter=json && codecov -f coverage/*.json",
     "fix": "gts fix",
-    "compile": "tsc -p .",
+    "compile": "tsc --target ES5 --module CommonJS",
```
To remove `allowSyntheticDefaultImports` option I directly transpiled typescript files to ES5 and CommonJS for running tests in node.

AndyOGo replied:
To remove `allowSyntheticDefaultImports` option I directly transpiled typescript files to ES5 and CommonJS for running tests in node.

AndyOGo commented on package.json:
```diff
@@ -16,10 +19,11 @@
     "fix": "gts fix",
-    "compile": "tsc -p .",
+    "compile": "tsc --target ES5 --module CommonJS",
+    "build-web": "microbundle",
```
`microbundle` uses the same `tsconfig.json` file now, expect `outDir` is set by `package.json` **main**, **module** and **unpkg** fields

AndyOGo replied:
`microbundle` uses the same `tsconfig.json` file now, expect `outDir` is set by `package.json` **main**, **module** and **unpkg** fields

AndyOGo commented on package.json:
```diff
@@ -47,17 +50,20 @@
     "codecov": "^3.5.0",
     "gts": "^2.0.0",
     "js-green-licenses": "^1.0.0",
+    "json-schema": "^0.2.5",
```
CI failed without this dev dependencies 🤔 

> Error: Failed to load plugin '@typescript-eslint' declared in '.eslintrc.json » ./node_modules/gts#overrides[0]': Cannot find module 'json-schema'

AndyOGo commented on tsconfig.json:
```diff
@@ -2,7 +2,9 @@
     "rootDir": ".",
-    "outDir": "build"
+    "outDir": "build",
+    "moduleResolution": "node",
```
Needed to find `axios` and `nock`

AndyOGo commented:
Hello @JustinBeckwith
Thank you again for your great review 👍 
Did you have a chance to look at my changes regarding your feedback?
Hope you like them:)
Enjoy your day

JustinBeckwith commented:
This is looking great!  A few quick things:
- Let's get the conflict on package.json sorted out
- If you could, please do a `rm -rf node_modules package-lock.json`, re-install, and then you should be able to drop the `json-schema` dependency.

AndyOGo commented:
Great, thank you @JustinBeckwith 
I rebased onto your `master` and removed the `json-schema` dependency :)

AndyOGo commented:
Hey @JustinBeckwith 
When do you plan the next release?
